### PR TITLE
Allow specifying alias analysis while registering new ops

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77690

### CODE REVIEW

File: torch/library.py [hunk 1]
```diff
@@ -39,8 +39,6 @@ def __repr__(self):
 
     def impl(self, op_name, fn, dispatch_key=''):
         if dispatch_key == '':
-            if self.dispatch_key == '':
```
Comment (anjali411) [inline torch/library.py:42]: seems like it's valid to pass no dispatch key

File: torch/csrc/utils/python_dispatch.cpp [hunk 2]
```diff
@@ -64,7 +64,7 @@ class PythonKernelHolder : public c10::OperatorKernel {
     py::gil_scoped_acquire g;
     auto args_kwargs = parseIValuesToPyArgsKwargs(op, arguments);
     auto obj = py::reinterpret_steal<py::object>(PyObject_Call(func_.ptr(getPyInterpreter()), args_kwargs.first.ptr(), args_kwargs.second.ptr()));
-    if (obj == nullptr) { throw python_error(); }
+    if (!obj) { throw python_error(); }
```
Comment (anjali411) [inline torch/csrc/utils/python_dispatch.cpp:67]: to make clangtidy happy

File: test/test_python_dispatch.py [hunk 3]
```diff
@@ -247,6 +247,34 @@ def my_sum_zt(*args, **kwargs):
...
+
+            assert "foo::_op" in str(_test.graph)
+            assert torch.ops.foo._op() is None
+            assert called[0] == 1
+
+            assert _test() is None
+            assert called[0] == 2
+
+        with self.assertRaises(AssertionError):
+            test_helper("")  # alias_analysis="FROM_SCHEMA"
```
Comment (anjali411) [inline test/test_python_dispatch.py:274]: fails at the first assert

**Review by ezyang — APPROVED** [review]

File: torch/library.py (hunk 1 above)
Comment (anjali411) [reply torch/library.py:42]: @bdhirsh shared that the default key is set to `CompositeImplicitAutograd` by default

**Comment by anjali411** [comment]
added another decorator to define and impl in the same line since you approved @ezyang

File: torch/library.py (hunk 1 above)
Comment (bdhirsh) [reply torch/library.py:42]: yeahh.... I dunno if we want the default to also be `CompositeImplicitAutograd` for python though.

The problem with that default for python is that the op will decompose (i.e. run the actual python code, calling the individual base ops) before hitting any transforms / backend code. What this means in practice is that `functionalize`, `__torch_dispatch__` and `AOTAUtograd` will never see the custom op that you registered - they'll just see the decomposed ops.

So I'd tentatively vote that if we want a default, it should be `CompositeExplicitAutograd`. We should probably shop that around with other people though.

**Comment by ezyang** [comment]
decorator seems fine

**Pull request closed without merging**